File: SmartFoodDeliveryApp/Services/FoodMemoryService.swift
import Foundation

@Observable
final class FoodMemoryService {
    private static let storageKey = "food_memory_v1"
    var memory: FoodMemory

    init() {
        if let data = UserDefaults.standard.data(forKey: Self.storageKey),
           let decoded = try? JSONDecoder().decode(FoodMemory.self, from: data) {
            memory = decoded
        } else {
            memory = FoodMemory()
        }
    }

    func save() {
        if let data = try? JSONEncoder().encode(memory) {
            UserDefaults.standard.set(data, forKey: Self.storageKey)
        }
    }

    func addDislikedIngredient(_ ingredient: String) {
        let normalized = ingredient.lowercased().trimmingCharacters(in: .whitespaces)
        guard !memory.dislikedIngredients.contains(normalized) else { return }
        memory.dislikedIngredients.append(normalized)
        save()
    }

    func removeDislikedIngredient(_ ingredient: String) {
        memory.dislikedIngredients.removeAll { $0 == ingredient.lowercased() }
        save()
    }

    func setSpiceLevel(_ level: SpiceLevel) {
        memory.spiceLevel = level
        save()
    }

    func setDefaultDrink(_ drink: String?) {
        memory.defaultDrink = drink
        save()
    }

    func setCommonNotes(_ notes: String?) {
        memory.commonNotes = notes
        save()
    }

    func contextString() -> String {
        var parts: [String] = []
        if !memory.dislikedIngredients.isEmpty {
            parts.append("Dislikes: \(memory.dislikedIngredients.joined(separator: ", "))")
        }
        parts.append("Spice level: \(memory.spiceLevel.rawValue)")
        if let drink = memory.defaultDrink {
            parts.append("Default drink: \(drink)")
        }
        if let notes = memory.commonNotes {
            parts.append("Usual notes: \(notes)")
        }
        if !memory.preferredCuisines.isEmpty {
            parts.append("Preferred cuisines: \(memory.preferredCuisines.joined(separator: ", "))")
        }
        return parts.isEmpty ? "No preferences saved yet." : parts.joined(separator: "\n")
    }
}

File: src/services/FoodMemoryService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  defaultFoodMemory,
  type FoodMemory,
  type SpiceLevel,
} from '../models/FoodMemory';

const STORAGE_KEY = 'food_memory_v1';

export class FoodMemoryService {
  memory: FoodMemory = defaultFoodMemory;

  async load(): Promise<void> {
    try {
      const raw = await AsyncStorage.getItem(STORAGE_KEY);
      if (!raw) {
        this.memory = defaultFoodMemory;
        return;
      }
      this.memory = { ...defaultFoodMemory, ...(JSON.parse(raw) as FoodMemory) };
    } catch {
      this.memory = defaultFoodMemory;
    }
  }

  private async save(): Promise<void> {
    try {
      await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(this.memory));
    } catch {
      // ignore for now
    }
  }

  async addDislikedIngredient(ingredient: string): Promise<void> {
    const normalized = ingredient.toLowerCase().trim();
    if (this.memory.dislikedIngredients.includes(normalized)) return;
    this.memory.dislikedIngredients.push(normalized);
    await this.save();
  }

  async removeDislikedIngredient(ingredient: string): Promise<void> {
    const normalized = ingredient.toLowerCase();
    this.memory.dislikedIngredients = this.memory.dislikedIngredients.filter(
      x => x !== normalized,
    );
    await this.save();
  }

  async setSpiceLevel(level: SpiceLevel): Promise<void> {
    this.memory.spiceLevel = level;
    await this.save();
  }

  async setDefaultDrink(drink: string | null): Promise<void> {
    this.memory.defaultDrink = drink ?? null;
    await this.save();
  }

  async setCommonNotes(notes: string | null): Promise<void> {
    this.memory.commonNotes = notes ?? null;
    await this.save();
  }

  contextString(): string {
    const parts: string[] = [];
    if (this.memory.dislikedIngredients.length > 0) {
      parts.push(
        `Dislikes: ${this.memory.dislikedIngredients.join(', ')}`,
      );
    }
    parts.push(`Spice level: ${this.memory.spiceLevel}`);
    if (this.memory.defaultDrink) {
      parts.push(`Default drink: ${this.memory.defaultDrink}`);
    }
    if (this.memory.commonNotes) {
      parts.push(`Usual notes: ${this.memory.commonNotes}`);
    }
    if (this.memory.preferredCuisines.length > 0) {
      parts.push(
        `Preferred cuisines: ${this.memory.preferredCuisines.join(', ')}`,
      );
    }
    return parts.length === 0
      ? 'No preferences saved yet.'
      : parts.join('\n');
  }
}

export const foodMemoryService = new FoodMemoryService();

